python/cuml/experimental/explainer/tree_shap.pyx

```diff
@@ -0,0 +1,93 @@
+#
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.common import input_to_cuml_array
+from cuml.common.array import CumlArray
+from cuml.internals import api_return_array
+from cuml.fil.fil import TreeliteModel
+
+from libcpp.memory cimport unique_ptr
+from libc.stdint cimport uintptr_t
+from libcpp.utility cimport move
+import numpy as np
+
+cdef extern from "treelite/c_api.h":
+    ctypedef void* ModelHandle
+    cdef int TreeliteQueryNumClass(ModelHandle handle, size_t* out)
+
+cdef extern from "treelite/c_api_common.h":
+    cdef const char* TreeliteGetLastError()
+
+cdef extern from "cuml/explainer/tree_shap.hpp" namespace "ML::Explainer":
+    cdef cppclass TreePathInfo:
+        pass
+
+    cdef unique_ptr[TreePathInfo] extract_path_info(ModelHandle model) except +
+    cdef void gpu_treeshap(const TreePathInfo* path_info,
+                           const float* data,
+                           size_t n_rows,
+                           size_t n_cols,
+                           float* out_preds) except +
+
+cdef class TreeExplainer_impl():
+    cdef ModelHandle model_ptr
+    cdef unique_ptr[TreePathInfo] path_info
+    cdef size_t num_class
+
+    def __cinit__(self, handle=None):
+        self.model_ptr = <ModelHandle> <uintptr_t> handle
+        self.num_class = 0
+        if TreeliteQueryNumClass(self.model_ptr, &self.num_class) != 0:
+            raise RuntimeError('Treelite error: {}'.format(
+                TreeliteGetLastError()))
+        self.path_info = move(extract_path_info(self.model_ptr))
+
+    def shap_values(self, X):
```

Comment:
Note that the shap package does not return the expected value column - the last column returned by GPUTreeShap which represents the model output when all features are turned off. Instead it sets this as the `self.expected_value` attribute of the TreeExplainer object: https://github.com/slundberg/shap/blob/429fb3e0ac2ef179f1cff7e1a64b4f7b26f41eb5/shap/explainers/_tree.py#L413.

I think we want to mimic this behaviour.

Reply:
Done